### PR TITLE
Encode Join Attributes in Bot Certificates

### LINKED ISSUE #49571: Workload ID: Embed selected Join Attributes in Bot Certificates
As per https://github.com/gravitational/teleport/pull/49133

Embed JoinAttributes in Bot Certificates and propagate to renewed or impersonated certificates.

### PR DESCRIPTION
**Pull request by strideynet**
As per RFD191: https://github.com/gravitational/teleport/pull/49133

Closes https://github.com/gravitational/teleport/issues/49571

Depends on https://github.com/gravitational/teleport/pull/50680

This PR encodes attributes produced by the join process into the certificates issued to Bots, so that these attributes can be used in authorization and templating decisions made when issuing Workload Identities.

In the v16/v17 backports, a disabled-by-default environment variable flag will be added. In a later release, we may enable this by default.

### CODE REVIEW

File: lib/tlsca/ca.go [hunk 1]
```diff
@@ -873,6 +883,19 @@ func (id *Identity) Subject() (pkix.Name, error) {
...
 
+	if id.JoinAttributes != nil {
+		encoded, err := protojson.Marshal(id.JoinAttributes)
+		if err != nil {
+			return pkix.Name{}, trace.Wrap(err, "encoding join attributes as protojson")
+		}
+		subject.ExtraNames = append(subject.ExtraNames,
+			pkix.AttributeTypeAndValue{
+				Type:  JoinAttributesASN1ExtensionOID,
+				Value: string(encoded),
```
Comment (strideynet) [inline lib/tlsca/ca.go:894]: Can/Should we store this as bytes instead?

File: lib/gitlab/gitlab.go [hunk 2]
```diff
@@ -129,3 +132,12 @@ func (c *IDTokenClaims) JoinAuditAttributes() (map[string]interface{}, error) {
 	}
 	return res, nil
 }
+
+func (c *IDTokenClaims) SetJoinAttrs(in *workloadidentityv1pb.JoinAttrs) proto.Message {
```
Comment (strideynet) [inline lib/gitlab/gitlab.go:136]: Need some opinions here - is it weird that this both mutates the passed in struct and returns the data it has injected? I'm not seeing too many other options but I'd like a second opinion in case I'm missing a simpler option.

Roughly - what I'm trying to achieve is:

- Inject the join metadata into workloadidentityv1pb.JoinAttrs for inclusion in the X509 cert as join attributes.
- Continue to be able to place this join metadata into the GoGo structpb.Struct in the audit event, and the Google structpb.Struct in the BotInstanceStatusAuthentication
- Avoid too much duplicated "per join method" code in the lib/auth/join.go handler. Ideally all the join methods can return an individual struct that implements a single method interface that can then be called by code further along to inject/extract this metadata.

Comment (timothyb89) [reply lib/gitlab/gitlab.go:136]: I guess at a minimum we could probably rename it to `SetAndReturnJoinAttrs()` or something so the behavior is at least not surprising.

That said, the thing that stands out to me is that we have 2 copies of each provider's JWT spec floating around now, one in proto form, and one as the `IDTokenClaims` for a particular provider, with copious amounts of copying between them. I'm guessing the idea is that our `IDTokenClaims` could be a different subset of the provider's claims than what Workload ID wants (or what we want to put in the audit log), but in practice I'm not sure if there's a real distinction between the two, or if there needs to be one?

In other words, could we make the proto definition the source of truth and eliminate `IDTokenClaims`? I think that would simplify the "convert, set, and return" logic to just "set", since we could just copy the full object into the `workloadidentityv1pb.JoinAttrs`. I think `oidc.IDToken.Claims()` deserialization would still work since they resulting `.pb.go` has `json` tags on the fields.

File: lib/tlsca/ca.go (hunk 1 above)
Comment (timothyb89) [reply lib/tlsca/ca.go:894]: I think the internal encoding implementation is identical (https://github.com/golang/go/blob/master/src/encoding/asn1/marshal.go#L41-L63) but I believe the values are explicitly typed. That doesn't matter much for us since we'll support whichever datatype we use, but I think tools (openssl, etc) can decide to render plaintext or hex depending on type. That is to say, I think strings are probably better due to the type that will be embedded alongside the value itself.

Technically I think ASN.1 has some native struct encoding we could use, and the go encoder supports it via field tags. That might be technically be "more" "correct" but probably adds too much complexity to be worth it.

File: lib/gitlab/gitlab.go (hunk 2 above)
Comment (strideynet) [reply lib/gitlab/gitlab.go:136]: > That said, the thing that stands out to me is that we have 2 copies of each provider's JWT spec floating around now, one in proto form, and one as the IDTokenClaims for a particular provider,

So I was mulling over this. It'd be certainly convenient if we could unmarshal directly into this struct, but, I think there's might be a few cases in which we'll want to convert a canonical field from an IdP into a more correct type.

If we take GitLab as an example, they have "ref_protected" and "environment_protected" claims which store a boolean value in string form (e.g "true" or "false"). This wasn't too problematic in our initial use-case where we've been storing these values for the purpose of auditing, but, as we start to look at rule writing/expressions/templating interacting with this data, then, us being able to convert these types for the user so that expressions can be more sane (e.g `join.gitlab.environment_protected && join.gitlab.ref_protected` rather than `join.gitlab.environment_protected == "true" && join.gitlab.ref_protected == "true"`) seems more valuable.

I suppose it's worth considering making `IDTokenClaims` unexported and the validator directly returning a `workloadidentityv1pb.GItLabJoinAttrs`. That would certainly "simplify" - and - in some regards, IDTokenClaims then just becomes an internal detail around how the JWT is unmarshaled and would effectively "not exist".

File: lib/tlsca/ca.go (hunk 1 above)
Comment (strideynet) [reply lib/tlsca/ca.go:894]: Yeah - I have a feeling that encoding as a string may just be the best bet. I think my biggest fear here is overly inflating certificate sizes - I'm yet to really determine the point at which it would become problematic. If testing does indicate this, then maybe we'll need to switch to encoding as binary rather than protojson.

Comment (strideynet) [reply lib/tlsca/ca.go:916]: After looking into this, it looks like storing Proto bytes instead actually causes some issues in the Subject DN due to the way Go's implementation of TLS handles parsing X509 certs. So we'll have to stick with string for now.

File: lib/spacelift/spacelift.go [hunk 3]
```diff
@@ -49,6 +51,21 @@ type IDTokenClaims struct {
...
+		CallerId:   c.CallerID,
+		RunType:    c.RunType,
+		RunId:      c.RunID,
+		Scope:      c.Scope,
+	}
+}
+
 // JoinAuditAttributes returns a series of attributes that can be inserted into
 // audit events related to a specific join.
 func (c *IDTokenClaims) JoinAuditAttributes() (map[string]interface{}, error) {
```
Comment (timothyb89) [inline lib/spacelift/spacelift.go:71]: Is `JoinAuditAttributes` still needed?

File: lib/githubactions/githubactions.go [hunk 4]
```diff
@@ -118,3 +120,24 @@ func (c *IDTokenClaims) JoinAuditAttributes() (map[string]interface{}, error) {
 	}
 	return res, nil
 }
```
Comment (timothyb89) [inline lib/githubactions/githubactions.go:122]: Awkward comment placement but as in spacelift, is `JoinAuditAttributes` still used?

**Review by timothyb89 — APPROVED** [review]
Looks good, and join attributes are showing up nicely in certs for me.

File: lib/spacelift/spacelift.go (hunk 3 above)
Comment (strideynet) [reply lib/spacelift/spacelift.go]: ab2adae788

File: lib/githubactions/githubactions.go (hunk 4 above)
Comment (strideynet) [reply lib/githubactions/githubactions.go]: ab2adae788

**Review by ravicious — APPROVED** [review]

File: lib/auth/join.go [hunk 5]
```diff
@@ -404,22 +427,13 @@ func (a *Server) generateCertsBot(
...
-		joinEvent.Attributes, err = apievents.EncodeMap(attributes)
-		if err != nil {
-			a.logger.WarnContext(ctx, "Unable to encode join attributes for audit event", "error", err)
-		}
-
-		auth.Metadata, err = structpb.NewStruct(attributes)
-		if err != nil {
-			a.logger.WarnContext(ctx, "Unable to encode struct value for join metadata", "error", err)
-		}
+	// TODO(noah): In v18, we can drop writing to the deprecated Metadata field.
```
[the remark this answers is not in the document]
Comment (strideynet) [reply lib/auth/join.go:433]: Ah yes - you are right. I think I wrote some of this code a long time ago 😅

[the remark this answers is not in the document]
Comment (strideynet) [reply lib/auth/join.go]: 4298e16f1b

File: lib/auth/join_iam.go [hunk 6]
```diff
@@ -260,41 +273,41 @@ func checkIAMAllowRules(identity *awsIdentity, token string, allowRules []*types
...
 	identity, err := executeSTSIdentityRequest(ctx, a.httpClientForAWSSTS, identityRequest)
 	if err != nil {
-		return trace.Wrap(err, "executing STS request")
+		return nil, trace.Wrap(err, "executing STS request")
 	}
 
 	// check that the node identity matches an allow rule for this token
 	if err := checkIAMAllowRules(identity, provisionToken.GetName(), provisionToken.GetAllowRules()); err != nil {
-		return trace.Wrap(err, "checking allow rules")
+		return identity, trace.Wrap(err, "checking allow rules")
```
[the remark this answers is not in the document]
Comment (strideynet) [reply lib/auth/join_iam.go:310]: 4298e16f1b

File: lib/auth/join_iam.go [hunk 7]
```diff
@@ -260,41 +273,41 @@ func checkIAMAllowRules(identity *awsIdentity, token string, allowRules []*types
 
 // checkIAMRequest checks if the given request satisfies the token rules and
 // included the required challenge.
-func (a *Server) checkIAMRequest(ctx context.Context, challenge string, req *proto.RegisterUsingIAMMethodRequest, cfg *iamRegisterConfig) error {
+func (a *Server) checkIAMRequest(ctx context.Context, challenge string, req *proto.RegisterUsingIAMMethodRequest, cfg *iamRegisterConfig) (*awsIdentity, error) {
```
[the remark this answers is not in the document]
Comment (strideynet) [reply lib/auth/join_iam.go:276]: It's so the information about the IAM entity that is joining can be included in audit logs. The audit log is emitted in a higher scope, but the actual decision about the entity and whether or not it can join, is made here.

[the remark this answers is not in the document]
Comment (strideynet) [reply lib/auth/join_iam.go:276]: I'll add a godoc.

[the remark this answers is not in the document]
Comment (strideynet) [reply lib/auth/join_iam.go:280]: 301e1f4cbc
